packages/monitor-v2/src/monitor-og/oSnapAutomation.ts

```diff
@@ -0,0 +1,393 @@
+// Snapshot proposal is flattened into multiple SnapshotProposalExpanded objects. Each SnapshotProposalExpanded object
+// contains one safe from the original Snapshot proposal together with all other properties from the original Snapshot
+// proposal.
+export interface SnapshotProposalExpanded extends Omit<SnapshotProposalGraphql, "plugins"> {
+  safe: SafeSnapSafe;
+}
+
+// Checks that currency is among supportedBonds and that the bond amount exactly matches.
+const isBondSupported = (currency: string, bond: string, supportedBonds?: SupportedBonds): boolean => {
+  for (const supportedCurrency in supportedBonds) {
+    if (ethersUtils.getAddress(currency) === ethersUtils.getAddress(supportedCurrency)) {
+      return supportedBonds[supportedCurrency] === bond;
+    }
+  }
+  return false;
+};
+
+// Filters through all monitored OGs and returns all supported modules with their parameters. Specifically, this checks
+// that standard parsable rules are present and that the bond currency and amount is supported.
+const getSupportedModules = async (params: MonitoringParams): Promise<SupportedModules> => {
+  const supportedModules: SupportedModules = {};
+
+  await Promise.all(
+    params.ogAddresses.map(async (ogAddress) => {
+      const og = await getOgByAddress(params, ogAddress);
+      const rules = await og.rules();
+      const parsedRules = parseRules(rules);
+      const currency = await og.collateral();
+      const bond = (await og.bondAmount()).toString();
+      if (parsedRules !== null && isBondSupported(currency, bond, params.supportedBonds))
+        supportedModules[ogAddress] = { parsedRules, currency, bond };
+    })
+  );
+
+  return supportedModules;
+};
+
+const getModuleParameters = (ogAddress: string, supportedModules: SupportedModules): SupportedParameters => {
+  return supportedModules[ethersUtils.getAddress(ogAddress)];
+};
+
+// Queries snapshot for all space proposals that have been closed and have a plugin of safeSnap. The query also filters
+// only for basic type proposals that oSnap automation supports. This uses provided retry config, but ultimately throws
+// if the Snapshot query fails after all retries.
+const getSnapshotProposals = async (
+  spaceId: string,
+  url: string,
+  retryOptions: RetryOptions
+): Promise<Array<SnapshotProposalGraphql>> => {
+  const query = gql(/* GraphQL */ `
+    query GetProposals($spaceId: String) {
+      proposals(
+        where: { space: $spaceId, type: "basic", plugins_contains: "safeSnap", scores_state: "final", state: "closed" }
+        orderBy: "created"
+        orderDirection: desc
+      ) {
+        id
+        ipfs
+        type
+        choices
+        start
+        end
+        state
+        space {
+          id
+        }
+        scores
+        quorum
+        scores_total
+        plugins
+      }
+    }
+  `);
+  const graphqlData = await retry(
```

Comment:
What happens if the request fails all the times (e.g. something wrong with the snapshot endpoint) and graphql data is empty. Do we handle this?